File: DsDotNet/nuget/Common/Dual.Common.Core.FS/TypeExtensions/Em.Dictionary.fs
namespace Dual.Common.Core.FS

open System.Runtime.CompilerServices
open System.Collections.Generic

type DictionaryExt =

    /// collection 과 keySelector, valueSelector 로부터 dictionary 생성
    [<Extension>]
    static member ToDictionary<'x, 'k, 'v when 'k : equality>(
        xs: 'x seq,
        keySelector: 'x -> 'k,
        valueSelector: 'x -> 'v
    ) =
        xs
        |> map (fun x -> keySelector x, valueSelector x)
        |> Tuple.toDictionary

    /// (Key, Value) tuple 로부터 dictionary 생성
    [<Extension>] static member ToDictionary<'k, 'v when 'k : equality>(kvs:('k*'v) seq) = Tuple.toDictionary kvs
    /// KeyValuePair 로부터 dictionary 생성
    [<Extension>] static member ToDictionary<'k, 'v when 'k : equality>(kvs:KeyValuePair<'k, 'v> seq) = kvs |> map (fun kv ->  kv.Key, kv.Value) |> Tuple.toDictionary

File: DsDotNet/nuget/Common/Dual.Common.Core/ExtensionMethods/EmDictionary.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;

namespace Dual.Common.Core
{
    public static class EmDictionary
    {
        /// <summary>
        /// K->V dictionary 를 V->K dictionary 로 변환
        /// </summary>
        public static Dictionary<V,K> SwapKeyValue<K,V>(this Dictionary<K,V> dict)
        {
            if (dict.Values.Count != dict.Values.Distinct().Count())
                throw new Exception($"Values are not distinct!");

            return dict.ToDictionary(kv => kv.Value, kv => kv.Key);
        }

        /// <summary>
        /// K->V ConcurrentDictionary 를 V->K dictionary 로 변환
        /// </summary>
        public static Dictionary<V, K> SwapKeyValue<K, V>(this ConcurrentDictionary<K, V> dict)
        {
            if (dict.Values.Count != dict.Values.Distinct().Count())
                throw new Exception($"Values are not distinct!");

            return dict.ToDictionary(kv => kv.Value, kv => kv.Key);
        }

    }
}
